### ISSUE TITLE
Compact Blocks

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#522]
Tracking subtasks here - 
- [x] short ids - #632
- [x] CompactBlock model with vecs of short ids for inputs|output|kernels - #637 
  - [x] Note: we want coinbase outputs/kernels to remain "full" (we never see these in advance) - #641 
  - [x] limit # coinbase outputs in a tx? #644
  - [x] only include kernel short_ids in comact blocks #668 
- [x] API support for retrieving a compact block based on height/hash (for debugging etc.) - #637
- [x] modify p2p block propagation to broadcast "header first" - #654
  - [x] ask for full block after receiving a header
  - [x] ask for compact block first
- [x] p2p support for requesting full block if we cannot hydrate compact block successfully (i.e. we are missing some inputs/outputs/kernels from the underlying txs)
- [x] tx pool (or otherwise) support for querying what we have available (and what we are missing) by short id(s) #645
- [x] functionality to take a compact block and hydrate it into a full block as necessary

---------

Rough notes, may be wildly inaccurate.

* Nodes already have the txs in their tx pools - compact blocks avoid sending this data around a 2nd time as part of the block.
* Compact blocks _may_ remove the need for "header first propagation" mimblewimble/grin/issues/257 (full blocks vs. compact blocks vs. block headers)
* non-existence of txs (in the pool?) is a complicating factor
* cut-through in the tx pool?

-----------

Notes on tx pool and input/output/kernel short_ids - 

We maintain a map of transactions by hash in the pool.
If a node receive a compact_block it needs to go to the tx pool and look for all inputs/outputs/kernels based on short_ids.
But we store them as txs, not granular as inputs/outputs/kernels.
So - add an additional map that indexes tx hashes by short_ids of all inputs/outputs/kernels?
This would allow us to retrieve all txs - and then we can build the block from these?

**Comment by sesam** [comment]
In this github repo, searching /compact.*block/ mostly gave core/src/core/block.rs that has functions compactable_block() and compact(&self). Grin maillist also not so much, so probably this was debated on other maillists and/or reddit.

**Comment by ignopeverell** [comment]
See here:

https://lists.launchpad.net/mimblewimble/msg00039.html

Check all the way to the end as there is a vulnerability in the initial scheme that I discovered later on (https://lists.launchpad.net/mimblewimble/msg00111.html). Note that any short identifier used for inputs/outputs/kernels need to be matched with its equivalent in the transaction pool.

**Comment by antiochp** [comment]
Reading through that now. I'll need to do a couple of passes before I start to understand everything in there...

**Comment by antiochp** [comment]
Captured the email chain from the mailing list here for reference - 

----------

In this discussion, I'm only interested in the size of a block when sent across the wire during propagation. It's desirable in this context to have small block sizes (in bytes) so that blocks can traverse the network quickly without causing too much additional traffic. As most nodes will generally receive the data included in a transaction twice (as the transaction is propagated and in a block), some obvious optimizations are possible.

As a reminder, a Grin block is composed of the following:

- A header, including the proof of work, some Merkle-ish roots, a reference to the previous block hash, etc. All the header is always required.

- A set of inputs as a direct link to previous outputs.

- A set of outputs, containing a commitment and a range proof.

- Transaction kernels, containing a signature, a public key and the explicit fee.


The largest piece of data by a good margin right now is the range proof (~5 KB each). Therefore we can introduce a first optimization:

- Block gets propagated sans range proof.

- Upon reception, a node checks whether it has all the range proofs for all outputs included in the block.

- If any missing, the node requests them from the peer that sent the block.

- Validation can proceed, skipping the range proofs that were already known (assuming they were already validated by the transaction pool).


The second largest piece of data is composed of the transaction kernels (~100 bytes each). We can't fully avoid sending those as they're not "anchored" on anything, like the range proof is anchored on an output. However we could send a shortened version of their hash. In this scheme, when sent across the wire, a block would only contain the first 16 bytes (or maybe even 12) of the hash of the whole kernel. Upon reception, an exchange similar to the one for range proofs would occur. Note that this requires the pool maintains a mapping of those truncated hashes with the actual kernels.

Beyond this, optimizations become non-trivial. We do *not* want to introduce some transaction-level hash as transactions are disaggregated in blocks by design. We could introduce something like inverted bloom filters for inputs or outputs but those are still rather small (32 to 40 bytes each) and so the tradeoff of space saving over complexity doesn't seem worth it at this point.

Both optimizations add a couple protocol level messages (request/response for each data type) and some additional interfaces and logic for the existence checks on the transaction pool. That creates a little bit of additional complexity, but stays contained and fairly manageable. The first optimization is by far the most interesting. I'm thinking of the second one more as a "nice to have".

Does it all sound reasonable? Am I missing something?

--------------------

If you do pursue this path, then the hash-prefix serves merely as a kernel hint.
That is, the receiver will look up how many kernel she has matching the hint.
If 0 or more than 1, it will query peers for a list of all known
matching kernels.
Uhm, that should be: ask the sending peer to expand the hint to a full kernel.
If exactly 1, then it will use that, but if validation fails, it will
again have to go back
to querying peers for ALL the hints this time, so that it may learn of other
matching kernels that would validate. In this approach, even 12 bytes
is overkill,
and somewhere between  and 8 bytes suffices.

Btw, how should we encode inputs? There's two ways to get away with 8 bytes:

1) a 32 bit block number and 32 bit index in sorted outputs of that block
2) a single 64 bit index in list all of outputs, sorted first by block.

--------------------

Now that we're going down that path, I'm starting to wonder whether we should treat inputs, outputs and kernels all in that same way. It's definitely simpler.

Here's the proposal:

- Across the wire, blocks have the regular list of inputs, outputs and kernels. At the minimum, they must have one output and one kernel (coinbase).

- Blocks can also include the first n bytes of input/output/kernel hashes that are part of the block but not fully included.

- A node receiving a block with partial hashes checks them against its the transaction pool.

- Known input/output/kernels are added, missing ones (or collisions) are asked to the block sending node all in one batch.

- Once everything is known, the full block can be validated.


This still leaves the possibility to have full blocks, including all the data. In the general case the blocks should be very compact and requests for more data will only be commonplace for nodes newly started. It also gives an incentive to miners to have a fairly full transaction pool (to avoid further requests).

As far as how many bytes of the hash, as John mentioned, 8 bytes (a u64) seem to suffice. I believe the chances of collisions with a pool of 100,000 elements would still be less than 0.00000003%. 4 bytes would likely be too small, a transaction pool of the same size would have 69% of collisions.

--------------------

Definitely a fan of compact blocks to mitigate some impact of a faster block time. A few random thoughts below:

On range proof validation: Doing all we can to reduce duplicate range proof transmission seems like a big win. I prefer the amended version in the last email to the original proposal. Using a partial output rather than an output identifier introduces a fair bit of complexity.

As far as I can tell, the specific construction of the output hash's preimage has not been determined. If we move forward with this mechanism, it becomes critically important that the hash covers the range proof itself. A situation where the hash resolves ambiguously should be avoided at all costs.

On inputs being part of the scheme: Since an input is by itself little more than a hash, it seems like the gain there is likely fairly small. We have thus far avoided the need to uniquely identify inputs by anything other than what output it is spending. I think I'd prefer inputs to be in the block in their entirety.

This is a cool design! I think nailing down some of the specific aspects of the design will require a bit of protocol testing but this makes sense as a foundation to start with.

--------------------

It should be sufficient for the output and its rangeproof to be separately committed to the chain to prevent ambiguity. Committing to rangeproofs, which are witness data and can be ignored (at a trust tradeoff), will reduce flexibility.

This is a good point. I'm not opposed to the rangeproof and output having separate identifiers. In the context of the compact block discussion I wanted to emphasize the ability to unambiguously commit to a specific (output, range proof) pair to avoid a search problem that quickly becomes untenable, but if this comes in the form of two separate identifiers resolved independently, that should be fine.

Right. It might be better to to have two Merkle structures. I know Igno wanted only one, for simplicity, but I think it's easiest and most efficient for witness data to be in a parallel tree. So you have a tree with kernels, inputs, outputs, and in parallel you have a tree with the respective signatures, "input witnesses", and rangeproofs.

This makes things unambiguous: to find a rangeproof for an output, you look in the witness tree at the same position that you found the output in the data tree.

Input witnesses could be nothing, or maybe it makes sense for them to be full Merkle proofs that the referenced inputs are currently in the UTXO tree.

--------------------

I wanted to amend my proposal on compact blocks to refine further the "first n bytes of input/output/kernel hashes". Obviously this opens up a possibility of collision forgery, which can then increase network chattiness.

BIP152 introduced a good solution by introducing short transaction ids (which we can generalize to inputs/outputs/kernels):

https://github.com/bitcoin/bips/blob/master/bip-0152.mediawiki#Short_transaction_IDs

It does force a full re-hashing of the pool on each block but siphash 2-4 is fast and we don't need to introduce a new crypto primitive as siphash 2-4 is already used by Cuckoo Cycle.

I was also playing with the idea of committing to the short tx ids in the block header (instead of the tx hashes) to avoid network-level malleability or compact blocks but after checking with instagibbs on bitcoin-wizards (thanks!) that seems like a premature optimization right now. Bitcoin doesn't have such a thing and just plain ban on misbehavior seems to work well enough so far.

--------------------

Not only that, but Cuckoo Cycle proceeds with exactly the same steps:

"Short transaction IDs are used to represent a transaction without
sending a full 256-bit hash. They are calculated by:
1. single-SHA256 hashing the block header with the nonce appended (in
little-endian)
2. Running SipHash-2-4 with the input being the transaction ID and the
keys (k0/k1) set to the first two little-endian 64-bit integers from
the above hash, respectively.
3. Dropping the 2 most significant bytes from the SipHash output to
make it 6 bytes."

Cuckoo Cycle on 2^{k+1} nodes looks for a cycle on the graph of
short transaction IDs truncated to k bits, where edge e, for 0<=e<2^k,
connects the transaction IDs 2*e and 2*e+1..

**Comment by antiochp** [comment]
> As far as I can tell, the specific construction of the output hash's preimage has not been determined. If we move forward with this mechanism, it becomes critically important that the hash covers the range proof itself. A situation where the hash resolves ambiguously should be avoided at all costs.

Is this still applicable? Today we explicitly exclude the rangeproof from the output hash.

**Comment by antiochp** [comment]
Conversation above considers - 
1. focusing on outputs/rangeproofs as this would be biggest win, vs - 
1. treating inputs/outputs/kernels the same and using some kind of ids

Are we going with the 2nd approach here? (@ignopeverell proposal in above conversation).

**Comment by ignopeverell** [comment]
I think so, we might as well. But I don't pretend to have thought out all the possible gotchas :-)

**Comment by antiochp** [comment]
This point on short_ids was confusing me (I was not understanding why this was necessary) - 
> It does force a full re-hashing of the pool on each block

It just clicked (and I am putting it down here for reference) - 
The inputs/outputs/kernels in the tx pool do not yet have short_ids because they do not (yet) belong to a block.
So when a node receives a compact block we need to rehash all the entries in the tx pool based on their potential candidacy for being included in this new block.
The point is that another mining node just included some set of these in the new block. And given the new block header hash we can generate (block specific) short_ids for everything currently in the tx pool.
